### PR TITLE
[Fixes #481] remove add registry sidecar from docker-build tests

### LINKED ISSUE #481: docker-build test task is failing
# Expected Behavior

No failure :) 

# Actual Behavior

The test are sometime failing : 

`FAILED: docker-build task has failed to comeback properly`

The main issue seems to be with this : 

```
+ docker build --no-cache -f ./Dockerfile -t localhost:5000/nocode .
unable to resolve docker endpoint: open /certs/client/ca.pem: no such file or directory
```

# Steps to Reproduce the Problem

1. Check the log [here](https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_catalog/479/pull-tekton-catalog-integration-tests/1292831095561129984/build-log.txt) from this [issue](https://github.com/tektoncd/catalog/pull/479#issuecomment-671397322)
2.
3.

# Additional Info

### PR DESCRIPTION
**Pull request by popcor255**
# Changes

## Overview
* Removed docker-build from skip task
* Removed add_sidecar_registry from tests
* Added a deployment with an image registry instead of sidecar.

## Summary
The end to end test attaches a sidecar with a image registry for tasks to push into to. This task is optional and is encouraged to be used for testing. However, the function add_sidecar_registry sets the sidecar registry instead of appending it. Not using the tasks will fix the test https://github.com/tektoncd/catalog/issues/481

## TLDR;
Script sets sidecar equal to an object instead of appending it. So, Im not going to use it in my test.

### CODE REVIEW

File: test/e2e-common.sh [hunk 1]
```diff
@@ -27,8 +27,7 @@ source $(dirname $0)/../vendor/github.com/tektoncd/plumbing/scripts/e2e-tests.sh
 # from our tests withouth having to go to an external registry.
 function add_sidecar_registry() {
     cp ${1} ${TMPF}.read
-
-    cat ${TMPF}.read | python3 -c 'import yaml;f=open(0, encoding="utf-8"); data=yaml.load(f.read(), Loader=yaml.FullLoader);data["spec"]["sidecars"]=[{"image":"registry", "name": "registry"}];print(yaml.dump(data, default_flow_style=False));' > ${TMPF}
+    cat ${TMPF}.read | python3 -c 'import yaml;f=open(0, encoding="utf-8"); data=yaml.load(f.read(), Loader=yaml.FullLoader);sidecars=data["spec"].get("sidecars"); sidecars = [[], data["spec"].get("sidecars")][sidecars != None]; sidecars.append({"image":"registry", "name": "registry"}); data["spec"]["sidecars"]=sidecars;print(yaml.dump(data, default_flow_style=False));' > ${TMPF}
```
Comment (ImJasonH) [inline test/e2e-common.sh:30]: This is frankly indecipherable. Can we make this more readable somehow?

Comment (chmouel) [reply test/e2e-common.sh:30]: yes we can.. let's maybe add this fix and work on refactoring this?

Comment (popcor255) [reply test/e2e-common.sh:30]: I 110% percent agree. There is tool already designed to do this kind of substitution called yq. I don't know why there is an arbitrary line of python in the first place. However, I feel like this someone else has had to say something before. This had to be put there for some reason. idk, why tho...

**Comment by popcor255** [comment]
:facepalm: This why its one line...
```bash
++ python3 utils.py add_sidecar_to_yaml_file
python3: can't open file 'utils.py': [Errno 2] No such file or directory
```

**Comment by chmouel** [comment]
This sounds good, thanks for doign this refactoring, any chances if you can squash the commits ? (preferrably if you can separate the refactoring from the fix but that's not absolutely necessary). thanks!

**Comment by vinamra28** [comment]
probably fetching the latest commits from upstream and rebasing the PR would help fix the CI

File: test/utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+def add_sidecar_to_yaml_file():
+    # This function is expecting a tekton task from stdin
+    with sys.stdin as file:
+        # Load yaml file passed by stdin
+        payload = yaml.load(file, Loader=yaml.FullLoader)
+        # If sidecars doesnt have any entries create an array for payload manipulation
+        if 'sidecars' not in payload['spec']:
+            payload['spec']['sidecars'] = []
+        # Append the local image registry to the yaml file
+        payload['spec']['sidecars'].append({"image": "registry", "name": "registry"})
```
Comment (ImJasonH) [inline test/utils.py:14]: Can we just specify the registry in the original YAML, instead of injecting in this way?

File: test/e2e-common.sh [hunk 3]
```diff
@@ -43,7 +42,7 @@ function install_pipeline_crd() {
   if [[ -n ${RELEASE_YAML} ]];then
 	latestreleaseyaml=${RELEASE_YAML}
   else
-    latestreleaseyaml=$(curl -s https://api.github.com/repos/tektoncd/pipeline/releases|python -c "import sys, json;x=json.load(sys.stdin);ass=x[0]['assets'];print([ x['browser_download_url'] for x in ass if x['name'] == 'release.yaml'][0])")
```
Comment (ImJasonH) [inline test/e2e-common.sh:46]: Could this just default to https://storage.googleapis.com/tekton-releases/latest/release.yaml, instead of querying GitHub for the latest release?

Comment (popcor255) [reply test/e2e-common.sh:46]: :+1: nice catch

File: test/utils.py (hunk 2 above)
Comment (popcor255) [reply test/utils.py:14]: I agree that injecting sidecars is not a good idea. I also think it would be a mistake to change the yaml for the sake of passing a test. The local image registry was only there to help us test `docker push`. Now, honestly its hurting.  I would perfer to push to my docker hub account then add an ephemeral image registry that provides no purpose for the end-user. Also, there are about 13 other tasks that use the add registry sidecar. Ignoring this method in docker-build does not fix the overall problem that this function does perform it's expected behavior. 
```bash
    cat ${TMPF}.read | python3 -c 'import yaml;f=open(0, encoding="utf-8"); data=yaml.load(f.read(), Loader=yaml.FullLoader);data["spec"]["sidecars"]=[{"image":"registry", "name": "registry"}];print(yaml.dump(data, default_flow_style=False));' > ${TMPF}
```
I do think having a namespace with a local image registry would be a good solution. Openshift provides an image registry that can be accessed in its own namespace. I think creating an uninsecured registry in its own namespace that gets cleaned up at the end of testing would be useful. Pods would be able to access it through it's svc. This would also help CI times cause we wouldn't have to spin up 14 image registry pods.

Comment (ImJasonH) [reply test/utils.py:14]: Oh I see what you're saying, the registry shouldn't be specified in the actual catalog task. That definitely makes sense, I wasn't paying enough attention. 🤦 

+1 to having e2e tests start by running a registry, run all the relevant tasks against it, and then delete the registry. This would let us test the tasks without modifying them, which might also hide breakages.

Comment (popcor255) [reply test/utils.py:14]: The only thing left that needs to be decided is a standard for naming and tagging images with a shared registry. (I.E we cant have 10 tasks build and push foo:latests to the registry without breaking the test).

Comment (popcor255) [reply test/utils.py:14]: Since, tasks already have a version naming and tagging it with those labels would work.
example:
<task-name>:<task-version>
localhost:5000/docker-build:0.1

**Comment by popcor255** [comment]
@ImJasonH @afrittoli @chmouel YAY! YAML! :tada: (could you take another look) :wink:

**Comment by popcor255** [comment]
how do i remove hold :crying_cat_face:

File: task/docker-build/0.1/docker-build.yaml [hunk 4]
```diff
@@ -34,6 +34,21 @@ spec:
   - name: push_extra_args
     description: Extra parameters passed for the push command when pushing images.
     default: ""
+  - name: docker_host
```
[the remark this answers is not in the document]
Comment (popcor255) [reply task/docker-build/0.1/docker-build.yaml]: Oh, big derp. That makes alot of sense because changing the parameters like I did, required me to change the docs anyway. This way is must better. +1

**Comment by chmouel** [comment]
/lgtm
/approve

Let's see how it goes!
